pkg/controllers/scheduling/schedule.go

```diff
@@ -0,0 +1,160 @@
+// cluster decisions. Empty cluster decisions will be appended if the number of selected
+// clusters is less than the number of desired clusters.
+func selectClusters(placement *clusterapiv1alpha1.Placement, clusters []*clusterapiv1.ManagedCluster) []clusterapiv1alpha1.ClusterDecision {
+	// truncate the cluster slice if the desired number of decisions is less than
+	// the number of the candidate clusters
+	numOfDecisions := len(clusters)
+	if placement.Spec.NumberOfClusters != nil {
+		numOfDecisions = int(*placement.Spec.NumberOfClusters)
+	}
+	if numOfDecisions < len(clusters) {
+		clusters = clusters[:numOfDecisions]
+	}
+
+	decisions := []clusterapiv1alpha1.ClusterDecision{}
+	for _, cluster := range clusters {
+		decisions = append(decisions, clusterapiv1alpha1.ClusterDecision{
+			ClusterName: cluster.Name,
+		})
+	}
+
+	// append unscheduled decisions if the desired number of decisions is larger than
+	// the number of the candidate clusters
+	for count := len(clusters); count < numOfDecisions; count++ {
+		decisions = append(decisions, clusterapiv1alpha1.ClusterDecision{})
+	}
+
+	return decisions
+}
+
+// bind updates the cluster decisions in the status of the placementdecisions with the given
+// cluster decision slice. New placementdecision will be created if no one exists.
+func bind(
+	ctx context.Context,
+	placement *clusterapiv1alpha1.Placement,
+	clusterDecisions []clusterapiv1alpha1.ClusterDecision,
+	clusterClient clusterclient.Interface,
+	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister,
+) error {
+	// query placementdecisions with label selector
+	requirement, err := labels.NewRequirement(placementLabel, selection.Equals, []string{placement.Name})
+	if err != nil {
+		return err
+	}
+	labelSelector := labels.NewSelector().Add(*requirement)
+	placementDecisions, err := placementDecisionLister.PlacementDecisions(placement.Namespace).List(labelSelector)
+	if err != nil {
+		return err
+	}
+
+	// TODO: support multiple placementdecisions for a placement
+	var placementDecision *clusterapiv1alpha1.PlacementDecision
+	switch {
+	case len(placementDecisions) > 0:
+		placementDecision = placementDecisions[0]
+	default:
+		// create a placementdecision if not exists
+		owner := metav1.NewControllerRef(placement, clusterapiv1alpha1.GroupVersion.WithKind("Placement"))
+		placementDecision = &clusterapiv1alpha1.PlacementDecision{
+			ObjectMeta: metav1.ObjectMeta{
+				GenerateName: fmt.Sprintf("%s-", placement.Name),
+				Namespace:    placement.Namespace,
+				Labels: map[string]string{
+					placementLabel: placement.Name,
+				},
+				OwnerReferences: []metav1.OwnerReference{*owner},
+			},
+		}
+		placementDecision, err = clusterClient.ClusterV1alpha1().PlacementDecisions(placement.Namespace).Create(ctx, placementDecision, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	// sort by cluster name
+	sort.SliceStable(clusterDecisions, func(i, j int) bool {
```

Comment:
Because you truncate in selectClusters this can still be unstable in cases where you are selecting a subset of clusters.  Currently it will appear stable because `matchWithClusterPredicates` will go evaluate in order, but if you add pre-indexing or multi-threading to that, you can end up unstable

Reply:
I'll add a TODO to sort clusters with scores before selectClusters() is invoking. That will ensure the selected clusters are stable. A followup PR will add the score logic.

Reply:
> I'll add a TODO to sort clusters with scores before selectClusters() is invoking. That will ensure the selected clusters are stable. A followup PR will add the score logic.

ok, I can accept this for now, but it carries a significant risk